diff -rup norm/common/normSession.cpp norm-dtncl/common/normSession.cpp
--- norm/common/normSession.cpp	2008-10-09 12:51:05.000000000 -0400
+++ norm-dtncl/common/normSession.cpp	2008-12-02 15:11:51.000000000 -0500
@@ -1645,7 +1645,31 @@ void NormSession::RxSocketRecvHandler(Pr
             rxMeasureSeqPrev = seq;
             return;
 #endif // RX_MEASURE_ONLY 
-            HandleReceiveMessage(msg, false);
+            if (address.GetType() == msg.AccessAddress().GetType() == ProtoAddress::IPv4)
+            {
+                if (address.IsMulticast() ||
+                    (((const struct sockaddr_in&)address.GetSockAddr()).sin_addr.s_addr ==
+                     ((const struct sockaddr_in&)msg.AccessAddress().GetSockAddr()).sin_addr.s_addr))
+                {
+                    HandleReceiveMessage(msg, false);
+                }
+                else
+                {
+                    NormSession *other_session = session_mgr.FindSession(msg.AccessAddress());
+                    if (other_session) {
+                        DMSG(4, "NormSession::RxSocketRecvHandler() info: rerouting message\n");
+                        other_session->HandleReceiveMessage(msg, false);
+                    }
+                    else
+                    {
+                        DMSG(0, "NormSession::RxSocketRecvHandler() warning: message dropped\n");
+                    }
+                }
+            }
+            else
+            {
+                HandleReceiveMessage(msg, false);
+            }
             msgLength = NormMsg::MAX_SIZE;
         }
         else
@@ -4213,3 +4237,20 @@ void NormSessionMgr::DeleteSession(class
     }
 }  // end NormSessionMgr::DeleteSession()
 
+NormSession* NormSessionMgr::FindSession(const ProtoAddress &address)
+{
+    NormSession *session = top_session;
+    while (session)
+    {
+        if (((const struct sockaddr_in&)address.GetSockAddr()).sin_addr.s_addr ==
+            ((const struct sockaddr_in&)session->address.GetSockAddr()).sin_addr.s_addr)
+        {
+            return session;
+        }
+        else
+        {
+            session = session->next;
+        }
+    }
+    return session;
+}  // end NormSessionMgr::FindSession()
diff -rup norm/common/normSession.h norm-dtncl/common/normSession.h
--- norm/common/normSession.h	2008-10-09 12:51:05.000000000 -0400
+++ norm-dtncl/common/normSession.h	2008-12-02 15:13:33.000000000 -0500
@@ -91,6 +91,8 @@ class NormSessionMgr
         NormController*          controller;
         
         class NormSession*       top_session;  // top of NormSession list
+
+        NormSession* FindSession(const ProtoAddress &address);
               
 };  // end class NormSessionMgr
 
